src/v/io/cache.h

```diff
@@ -0,0 +1,648 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include <boost/intrusive/list.hpp>
+#include <boost/intrusive/parent_from_member.hpp>
+
+#include <algorithm>
+#include <optional>
+
+namespace testing_details {
+class cache_hook_accessor;
+}; // namespace testing_details
+
+/**
+ * \defgroup cache Cache
+ *
+ * An implementation of the s3-fifo cache eviction algorithm.
+ *
+ * Yang, et al., "FIFO queues are all you need for cache eviction", SOSP '23,
+ * https://dl.acm.org/doi/10.1145/3600006.3613147
+ *
+ * The s3-fifo eviction algorithm uses three queues named small, main, and
+ * ghost. The implementation of the s3-fifo algorithm provides a concrete
+ * container for the small and main queues, and the user implements the ghost
+ * queue and any required indexing.
```

Comment:
The motivation is that the ghost queue implementation can be an optimization. For example, the page cache uses a btree_map because we actually want ordering for our entries, but you could easily use a hash table. Both of those are associative container, so in principle we could make the container type a template. However, the ghost fifo queue could also be implemented as a literal fifo queue, which we'll likely do the open-file-handle cache which doesn't really need a look-up mechanism.

Reply:
@nvartolomei yep, added!